File: app/src/main/java/com/baijum/ukufretboard/ui/VoiceLeadingView.kt
package com.baijum.ukufretboard.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.baijum.ukufretboard.R
import com.baijum.ukufretboard.data.Notes
import com.baijum.ukufretboard.domain.ChordVoicing
import com.baijum.ukufretboard.domain.VoiceLeading
import com.baijum.ukufretboard.viewmodel.UkuleleString

/**
 * Dedicated view for exploring voice leading through a chord progression.
 *
 * Shows optimal voicing transitions step-by-step with:
 * - A chord timeline highlighting the current transition pair
 * - Side-by-side chord diagrams with common tone highlighting
 * - Per-string movement breakdown
 * - Audio playback for individual chords and the full progression
 *
 * @param path The computed optimal voice leading path.
 * @param tuning Current ukulele tuning (for string name labels).
 * @param onBack Callback to exit the voice leading view.
 * @param onPlayVoicing Callback to play a single voicing (null if sound disabled).
 * @param onPlayAll Callback to play all voicings in sequence (null if sound disabled).
 * @param leftHanded Whether to mirror diagrams for left-handed players.
 * @param modifier Optional modifier.
 */
@Composable
fun VoiceLeadingView(
    path: VoiceLeading.Path,
    tuning: List<UkuleleString>,
    onBack: () -> Unit,
    onPlayVoicing: ((ChordVoicing) -> Unit)?,
    onPlayAll: ((List<ChordVoicing>) -> Unit)?,
    leftHanded: Boolean = false,
    modifier: Modifier = Modifier,
) {
    var currentStep by remember { mutableIntStateOf(0) }
    val hasTransitions = path.transitions.isNotEmpty()

    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(top = 8.dp),
    ) {
        // Header
        VoiceLeadingHeader(
            path = path,
            onBack = onBack,
        )

        HorizontalDivider(modifier = Modifier.padding(vertical = 4.dp))

        // Chord timeline
        ChordTimeline(
            steps = path.steps,
            currentStep = currentStep,
            hasTransitions = hasTransitions,
            onStepSelected = { i ->
                if (i < path.transitions.size) {
                    currentStep = i
                } else if (i > 0) {
                    currentStep = i - 1
                }
            },
        )

        Spacer(modifier = Modifier.height(8.dp))

        // Main scrollable content
        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp),
        ) {
            if (hasTransitions) {
                val fromStep = path.steps[currentStep]
                val toStep = path.steps[currentStep + 1]
                val transition = path.transitions[currentStep]

                // Two diagrams side by side
                TransitionDiagrams(
                    fromStep = fromStep,
                    toStep = toStep,
                    transition = transition,
                    leftHanded = leftHanded,
                    onPlayFrom = onPlayVoicing?.let { play -> { play(fromStep.voicing) } },
                    onPlayTo = onPlayVoicing?.let { play -> { play(toStep.voicing) } },
                )

                Spacer(modifier = Modifier.height(16.dp))

                // Transition details card
                TransitionSummaryCard(
                    transition = transition,
                    fromVoicing = fromStep.voicing,
                    toVoicing = toStep.voicing,
                    tuning = tuning,
                )

                Spacer(modifier = Modifier.height(12.dp))

                // Total path summary
                TotalPathSummary(path = path)
            } else {
                // Single chord — no transitions to show
                Text(
                    text = stringResource(R.string.voice_leading_single_chord),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(vertical = 16.dp),
                )
                VerticalChordDiagram(
                    voicing = path.steps[0].voicing,
                    onClick = { onPlayVoicing?.invoke(path.steps[0].voicing) },
                    leftHanded = leftHanded,
                )
            }

            Spacer(modifier = Modifier.height(8.dp))
        }

        // Bottom navigation controls
        if (hasTransitions) {
            HorizontalDivider()
            BottomControls(
                currentStep = currentStep,
                totalTransitions = path.transitions.size,
                onPrev = { if (currentStep > 0) currentStep-- },
                onNext = { if (currentStep < path.transitions.size - 1) currentStep++ },
                onPlayAll = onPlayAll?.let { play ->
                    { play(path.steps.map { it.voicing }) }
                },
            )
        }
    }
}

// ── Header ──────────────────────────────────────────────────────────────

@Composable
private fun VoiceLeadingHeader(
    path: VoiceLeading.Path,
    onBack: () -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        IconButton(onClick = onBack) {
            Icon(
                imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                contentDescription = stringResource(R.string.cd_back_progressions),
            )
        }
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = path.progressionName,
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
            )
            Text(
                text = stringResource(R.string.voice_leading_title, Notes.enharmonicForKey(path.keyRoot, path.keyRoot)),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

// ── Chord Timeline ──────────────────────────────────────────────────────

/**
 * Horizontal row of chord names showing the full progression.
 * The current "from" chord has a primary background and the "to" chord
 * has a secondary background, making the active transition pair obvious.
 */
@Composable
private fun ChordTimeline(
    steps: List<VoiceLeading.Step>,
    currentStep: Int,
    hasTransitions: Boolean,
    onStepSelected: (Int) -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState())
            .padding(horizontal = 16.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        steps.forEachIndexed { i, step ->
            if (i > 0) {
                Text(
                    text = "\u2192", // →
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            val isFrom = hasTransitions && i == currentStep
            val isTo = hasTransitions && i == currentStep + 1

            Box(
                modifier = Modifier
                    .background(
                        color = when {
                            isFrom -> MaterialTheme.colorScheme.primaryContainer
                            isTo -> MaterialTheme.colorScheme.secondaryContainer
                            else -> MaterialTheme.colorScheme.surfaceVariant
                        },
                        shape = RoundedCornerShape(8.dp),
                    )
                    .clickable { onStepSelected(i) }
                    .padding(horizontal = 12.dp, vertical = 6.dp),
                contentAlignment = Alignment.Center,
            ) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(
                        text = step.numeral,
                        style = MaterialTheme.typography.labelSmall,
                        color = when {
                            isFrom -> MaterialTheme.colorScheme.onPrimaryContainer
                            isTo -> MaterialTheme.colorScheme.onSecondaryContainer
                            else -> MaterialTheme.colorScheme.onSurfaceVariant
                        },
                    )
                    Text(
                        text = step.chordName,
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = if (isFrom || isTo) FontWeight.Bold else FontWeight.Normal,
                        color = when {
                            isFrom -> MaterialTheme.colorScheme.onPrimaryContainer
                            isTo -> MaterialTheme.colorScheme.onSecondaryContainer
                            else -> MaterialTheme.colorScheme.onSurfaceVariant
                        },
                    )
                }
            }
        }
    }
}

// ── Transition Diagrams ─────────────────────────────────────────────────

/**
 * Two chord diagrams side by side with an arrow between them.
 * The "to" diagram highlights common tones in a distinct color.
 */
@Composable
private fun TransitionDiagrams(
    fromStep: VoiceLeading.Step,
    toStep: VoiceLeading.Step,
    transition: VoiceLeading.TransitionInfo,
    leftHanded: Boolean,
    onPlayFrom: (() -> Unit)?,
    onPlayTo: (() -> Unit)?,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceEvenly,
        verticalAlignment = Alignment.Top,
    ) {
        // "From" diagram
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier.weight(1f),
        ) {
            Text(
                text = fromStep.numeral,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            VerticalChordDiagram(
                voicing = fromStep.voicing,
                onClick = { onPlayFrom?.invoke() },
                leftHanded = leftHanded,
            )
            if (onPlayFrom != null) {
                IconButton(
                    onClick = onPlayFrom,
                    modifier = Modifier.size(32.dp),
                ) {
                    Icon(
                        imageVector = Icons.Filled.PlayArrow,
                        contentDescription = stringResource(R.string.cd_play_item, fromStep.chordName),
                        modifier = Modifier.size(20.dp),
                    )
                }
            }
        }

        // Arrow
        Icon(
            imageVector = Icons.AutoMirrored.Filled.ArrowForward,
            contentDescription = stringResource(R.string.cd_to),
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier
                .padding(top = 60.dp)
                .size(24.dp),
        )

        // "To" diagram with common tone highlighting
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier.weight(1f),
        ) {
            Text(
                text = toStep.numeral,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            VerticalChordDiagram(
                voicing = toStep.voicing,
                onClick = { onPlayTo?.invoke() },
                leftHanded = leftHanded,
                commonToneIndices = transition.commonToneIndices,
            )
            if (onPlayTo != null) {
                IconButton(
                    onClick = onPlayTo,
                    modifier = Modifier.size(32.dp),
                ) {
                    Icon(
                        imageVector = Icons.Filled.PlayArrow,
                        contentDescription = stringResource(R.string.cd_play_item, toStep.chordName),
                        modifier = Modifier.size(20.dp),
                    )
                }
            }
        }
    }
}

// ── Transition Summary ──────────────────────────────────────────────────

/**
 * Card showing per-string movement details for a transition.
 * Common tones are highlighted with a colored dot and descriptive text.
 * Moving fingers show the fret change with directional arrows.
 */
@Composable
private fun TransitionSummaryCard(
    transition: VoiceLeading.TransitionInfo,
    fromVoicing: ChordVoicing,
    toVoicing: ChordVoicing,
    tuning: List<UkuleleString>,
) {
    val commonCount = transition.commonToneIndices.size
    val commonLabel = if (commonCount != 1)
        stringResource(R.string.voice_leading_common_tones_plural, commonCount)
    else
        stringResource(R.string.voice_leading_common_tones, commonCount)
    val distLabel = if (transition.totalDistance != 1)
        stringResource(R.string.voice_leading_frets_movement, transition.totalDistance)
    else
        stringResource(R.string.voice_leading_fret_movement, transition.totalDistance)

    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant,
        ),
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            // Summary headline
            Text(
                text = "$commonLabel \u00B7 $distLabel",
                style = MaterialTheme.typography.titleSmall,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurface,
            )

            Spacer(modifier = Modifier.height(12.dp))

            // Per-string breakdown
            tuning.forEachIndexed { i, string ->
                val fromFret = fromVoicing.frets[i]
                val toFret = toVoicing.frets[i]
                val movement = transition.movements[i]
                val isCommon = i in transition.commonToneIndices

                val fromNoteName = if (fromFret == ChordVoicing.MUTED) "x"
                    else Notes.pitchClassToName(
                        (string.openPitchClass + fromFret) % Notes.PITCH_CLASS_COUNT
                    )
                val toNoteName = if (toFret == ChordVoicing.MUTED) "x"
                    else Notes.pitchClassToName(
                        (string.openPitchClass + toFret) % Notes.PITCH_CLASS_COUNT
                    )

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 2.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    // String name label
                    Text(
                        text = "${string.name}:",
                        style = MaterialTheme.typography.bodySmall,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.width(24.dp),
                    )

                    if (isCommon) {
                        // Common tone — finger stays put
                        Box(
                            modifier = Modifier
                                .size(8.dp)
                                .background(
                                    MaterialTheme.colorScheme.secondary,
                                    CircleShape,
                                ),
                        )
                        Spacer(modifier = Modifier.width(6.dp))
                        val fretLabel = when (fromFret) {
                            ChordVoicing.MUTED -> stringResource(R.string.voice_leading_muted)
                            0 -> stringResource(R.string.label_open)
                            else -> stringResource(R.string.voice_leading_fret_prefix) + "$fromFret"
                        }
                        Text(
                            text = stringResource(R.string.voice_leading_stays) + "$fretLabel ($fromNoteName)",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.secondary,
                        )
                    } else {
                        // Moving finger
                        val fromLabel = when (fromFret) {
                            ChordVoicing.MUTED -> stringResource(R.string.voice_leading_muted)
                            0 -> stringResource(R.string.label_open)
                            else -> stringResource(R.string.voice_leading_fret_prefix) + "$fromFret"
                        }
                        val toLabel = when (toFret) {
                            ChordVoicing.MUTED -> stringResource(R.string.voice_leading_muted)
                            0 -> stringResource(R.string.label_open)
                            else -> stringResource(R.string.voice_leading_fret_prefix) + "$toFret"
                        }
                        val arrow = when {
                            movement > 0 -> "\u2191$movement"  // ↑
                            movement < 0 -> "\u2193${-movement}" // ↓
                            else -> ""
                        }
                        Text(
                            text = "$fromLabel ($fromNoteName)" + stringResource(R.string.voice_leading_arrow) + "$toLabel ($toNoteName)  $arrow",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
            }

            // Educational tip
            if (transition.commonToneIndices.isNotEmpty()) {
                Spacer(modifier = Modifier.height(12.dp))
                val commonStrings = transition.commonToneIndices
                    .sorted()
                    .joinToString(", ") { tuning[it].name }
                val plural = transition.commonToneIndices.size > 1
                Text(
                    text = if (plural)
                        stringResource(R.string.voice_leading_tip_plural, commonStrings)
                    else
                        stringResource(R.string.voice_leading_tip_singular, commonStrings),
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.Medium,
                    color = MaterialTheme.colorScheme.primary,
                )
            }
        }
    }
}

// ── Total Path Summary ──────────────────────────────────────────────────

@Composable
private fun TotalPathSummary(path: VoiceLeading.Path) {
    val transCount = path.transitions.size
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surface,
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = 0.dp),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(
                text = stringResource(R.string.voice_leading_total_path, path.totalDistance, transCount),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
            )
        }
    }
}

// ── Bottom Controls ─────────────────────────────────────────────────────

@Composable
private fun BottomControls(
    currentStep: Int,
    totalTransitions: Int,
    onPrev: () -> Unit,
    onNext: () -> Unit,
    onPlayAll: (() -> Unit)?,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceEvenly,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            OutlinedButton(
                onClick = onPrev,
                enabled = currentStep > 0,
            ) {
                Text("\u25C0 " + stringResource(R.string.voice_leading_prev))
            }

            if (onPlayAll != null) {
                Button(onClick = onPlayAll) {
                    Icon(
                        imageVector = Icons.Filled.PlayArrow,
                        contentDescription = stringResource(R.string.cd_play_all_voice_leadings),
                        modifier = Modifier.size(18.dp),
                    )
                    Spacer(modifier = Modifier.width(4.dp))
                    Text(stringResource(R.string.voice_leading_play_all))
                }
            }

            OutlinedButton(
                onClick = onNext,
                enabled = currentStep < totalTransitions - 1,
            ) {
                Text(stringResource(R.string.voice_leading_next) + " \u25B6")
            }
        }

        Spacer(modifier = Modifier.height(4.dp))

        Text(
            text = stringResource(R.string.voice_leading_step, currentStep + 1, totalTransitions),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

File: iosApp/UkuleleCompanion/Views/VoiceLeadingView.swift
import SwiftUI
import shared

struct VoiceLeadingView: View {
    let progression: Progression
    let keyRoot: Int32

    @State private var currentStep = 0
    @State private var path: VoiceLeading.Path?

    @StateObject private var tonePlayer = TonePlayer()
    // Voice-leading paths and note names follow the selected tuning (#576).
    // Left-handed mirroring now comes from the diagram's environment.
    private let tuning: [shared.UkuleleString] = FretboardPreferences.tuning.asUkuleleStrings

    var body: some View {
        Group {
            if let path = path {
                pathContent(path)
            } else {
                VStack(spacing: 16) {
                    Text("Unable to compute voice leading path.")
                        .foregroundStyle(.secondary)
                    Text("Some chords may not have playable voicings in this key.")
                        .font(.caption)
                        .foregroundStyle(.tertiary)
                }
                .padding()
            }
        }
        .navigationTitle("Voice Leading")
        .navigationBarTitleDisplayMode(.inline)
        .onAppear { computePath() }
    }

    private func computePath() {
        path = VoiceLeading.shared.computeOptimalPath(
            progression: progression,
            keyRoot: keyRoot,
            tuning: tuning
        )
    }

    @ViewBuilder
    private func pathContent(_ path: VoiceLeading.Path) -> some View {
        let steps = path.steps.asArray(of: VoiceLeading.Step.self)
        let transitions = path.transitions.asArray(of: VoiceLeading.TransitionInfo.self)
        let hasTransitions = !transitions.isEmpty

        VStack(spacing: 0) {
            // Header info
            VStack(spacing: 4) {
                Text(path.progressionName)
                    .font(.headline)
                Text("Key of \(Notes.shared.enharmonicForKey(pitchClass: path.keyRoot, keyRoot: nil, isMinor: false))")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            }
            .padding(.vertical, 8)

            Divider()

            // Chord timeline
            chordTimeline(steps: steps, hasTransitions: hasTransitions)
                .padding(.vertical, 8)

            // Main content
            ScrollView {
                VStack(spacing: 16) {
                    if hasTransitions {
                        let fromStep = steps[currentStep]
                        let toStep = steps[currentStep + 1]
                        let transition = transitions[currentStep]

                        transitionDiagrams(from: fromStep, to: toStep, transition: transition)
                        transitionSummaryCard(from: fromStep, to: toStep, transition: transition)
                        totalPathSummary(path: path)
                    } else {
                        Text("Single chord - no transitions to show.")
                            .font(.subheadline)
                            .foregroundStyle(.secondary)
                            .padding(.vertical, 16)

                        if let voicing = steps.first?.voicing {
                            ChordDiagramView(voicing: voicing, chordName: steps[0].chordName)
                        }
                    }
                }
                .padding(.horizontal)
            }

            // Bottom controls
            if hasTransitions {
                Divider()
                bottomControls(totalTransitions: transitions.count)
            }
        }
    }

    // MARK: - Chord Timeline

    private func chordTimeline(steps: [VoiceLeading.Step], hasTransitions: Bool) -> some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 4) {
                ForEach(0..<steps.count, id: \.self) { i in
                    if i > 0 {
                        Text("\u{2192}")
                            .font(.caption2)
                            .foregroundStyle(.secondary)
                    }

                    let isFrom = hasTransitions && i == currentStep
                    let isTo = hasTransitions && i == currentStep + 1

                    Button {
                        let transCount = (self.path?.transitions as? [Any])?.count ?? 0
                        if i < transCount {
                            currentStep = i
                        } else if i > 0 {
                            currentStep = i - 1
                        }
                    } label: {
                        VStack(spacing: 2) {
                            Text(steps[i].numeral)
                                .font(.caption2)
                            Text(steps[i].chordName)
                                .font(.caption.bold())
                        }
                        .padding(.horizontal, 10)
                        .padding(.vertical, 6)
                        .background(
                            isFrom ? Color.accentColor :
                            isTo ? Color.accentColor.opacity(0.3) :
                            Color(.systemGray5)
                        )
                        .foregroundStyle(isFrom ? .white : .primary)
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                    .buttonStyle(.plain)
                    .accessibilityCombined(
                        label: "\(steps[i].numeral) \(steps[i].chordName)",
                        value: isFrom ? "Current from" : isTo ? "Current to" : nil
                    )
                }
            }
            .padding(.horizontal)
        }
    }

    // MARK: - Transition Diagrams

    private func transitionDiagrams(
        from fromStep: VoiceLeading.Step,
        to toStep: VoiceLeading.Step,
        transition: VoiceLeading.TransitionInfo
    ) -> some View {
        let commonIndices = ((transition.commonToneIndices as? Set<KotlinInt>) ?? Set())
            .map { $0.intValue }
        let commonSet = Set(commonIndices)

        return HStack(alignment: .top) {
            // From diagram
            VStack(spacing: 4) {
                Text(fromStep.numeral)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                ChordDiagramView(voicing: fromStep.voicing, chordName: fromStep.chordName,
                                 commonToneIndices: commonSet)
                Button {
                    playVoicing(fromStep.voicing)
                } label: {
                    Image(systemName: "play.fill")
                        .font(.caption)
                }
                .accessibilityLabel("Play \(fromStep.chordName)")
            }
            .frame(maxWidth: .infinity)

            // Arrow
            Image(systemName: "arrow.right")
                .foregroundStyle(.secondary)
                .padding(.top, 60)
                .accessibilityHidden(true)

            // To diagram
            VStack(spacing: 4) {
                Text(toStep.numeral)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                ChordDiagramView(voicing: toStep.voicing, chordName: toStep.chordName,
                                 commonToneIndices: commonSet)
                Button {
                    playVoicing(toStep.voicing)
                } label: {
                    Image(systemName: "play.fill")
                        .font(.caption)
                }
                .accessibilityLabel("Play \(toStep.chordName)")
            }
            .frame(maxWidth: .infinity)
        }
    }

    // MARK: - Transition Summary Card

    private func transitionSummaryCard(
        from fromStep: VoiceLeading.Step,
        to toStep: VoiceLeading.Step,
        transition: VoiceLeading.TransitionInfo
    ) -> some View {
        let commonIndices = (transition.commonToneIndices as? Set<KotlinInt>) ?? Set()
        let commonCount = commonIndices.count
        let movements = transition.movements.asInts

        return VStack(alignment: .leading, spacing: 8) {
            let commonLabel = commonCount == 1 ? "1 common tone" : "\(commonCount) common tones"
            let distLabel = transition.totalDistance == 1
                ? "1 fret movement"
                : "\(transition.totalDistance) frets movement"

            Text("\(commonLabel) \u{00B7} \(distLabel)")
                .font(.subheadline.bold())

            let fromFrets = fromStep.voicing.fretInts
            let toFrets = toStep.voicing.fretInts

            ForEach(0..<tuning.count, id: \.self) { i in
                let fromFret = fromFrets[i]
                let toFret = toFrets[i]
                let movement = movements[i]
                let isCommon = commonIndices.contains(KotlinInt(int: Int32(i)))

                HStack(spacing: 4) {
                    Text("\(tuning[i].name):")
                        .font(.caption.bold())
                        .frame(width: 24, alignment: .leading)

                    if isCommon {
                        Circle()
                            .fill(Color.blue)
                            .frame(width: 8, height: 8)
                        let fretLabel = fromFret == 0 ? "open" : "fret \(fromFret)"
                        let noteName = noteName(stringIndex: i, fret: fromFret)
                        Text("stays \(fretLabel) (\(noteName))")
                            .font(.caption)
                            .foregroundStyle(.blue)
                    } else {
                        let fromLabel = fromFret == 0 ? "open" : "fret \(fromFret)"
                        let toLabel = toFret == 0 ? "open" : "fret \(toFret)"
                        let fromNote = noteName(stringIndex: i, fret: fromFret)
                        let toNote = noteName(stringIndex: i, fret: toFret)
                        let arrow = movement > 0 ? "\u{2191}\(movement)" : movement < 0 ? "\u{2193}\(-movement)" : ""
                        Text("\(fromLabel) (\(fromNote)) \u{2192} \(toLabel) (\(toNote))  \(arrow)")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }
            }

            if commonCount > 0 {
                let commonStrings = commonIndices.sorted { $0.int32Value < $1.int32Value }
                    .map { tuning[Int($0.int32Value)].name }
                    .joined(separator: ", ")
                let tip = commonCount == 1
                    ? "Keep your finger on string \(commonStrings) - it stays the same!"
                    : "Keep your fingers on strings \(commonStrings) - they stay the same!"
                Text(tip)
                    .font(.caption)
                    .foregroundStyle(Color.accentColor)
                    .padding(.top, 4)
            }
        }
        .padding()
        .background(Color(.secondarySystemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    // MARK: - Total Path Summary

    private func totalPathSummary(path: VoiceLeading.Path) -> some View {
        let transCount = path.transitions.count
        let transLabel = "\(transCount) transition\(transCount == 1 ? "" : "s")"
        return Text("Total path: \(path.totalDistance) frets across \(transLabel)")
            .font(.caption)
            .foregroundStyle(.secondary)
            .frame(maxWidth: .infinity)
            .padding()
            .background(Color(.systemBackground))
            .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    // MARK: - Bottom Controls

    private func bottomControls(totalTransitions: Int) -> some View {
        VStack(spacing: 4) {
            HStack {
                Button {
                    if currentStep > 0 { currentStep -= 1 }
                } label: {
                    Text("\u{25C0} Prev")
                }
                .disabled(currentStep == 0)

                Spacer()

                Button {
                    playAllVoicings()
                } label: {
                    Label("Play All", systemImage: "play.fill")
                }
                .buttonStyle(.borderedProminent)

                Spacer()

                Button {
                    if currentStep < totalTransitions - 1 { currentStep += 1 }
                } label: {
                    Text("Next \u{25B6}")
                }
                .disabled(currentStep >= totalTransitions - 1)
            }
            .padding(.horizontal)

            Text("\(currentStep + 1) of \(totalTransitions)")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .padding(.vertical, 8)
    }

    // MARK: - Audio

    private func playVoicing(_ voicing: ChordVoicing) {
        let frets = voicing.fretInts
        let pitchClasses = frets.enumerated().compactMap { i, fret -> Int32? in
            guard fret >= 0 else { return nil }
            return (tuning[i].openPitchClass + Int32(fret)) % 12
        }
        tonePlayer.playChord(pitchClasses: pitchClasses, strumDelayMs: 40)
    }

    private func playAllVoicings() {
        guard let path = path else { return }
        let steps = path.steps.asArray(of: VoiceLeading.Step.self)
        for (i, step) in steps.enumerated() {
            DispatchQueue.main.asyncAfter(deadline: .now() + Double(i) * 0.6) {
                self.playVoicing(step.voicing)
            }
        }
    }

    // MARK: - Helpers

    private func noteName(stringIndex: Int, fret: Int) -> String {
        if fret < 0 { return "x" }
        let pc = (tuning[stringIndex].openPitchClass + Int32(fret)) % 12
        return Notes.shared.pitchClassToName(pitchClass: pc)
    }
}
